See query details in `README.md`.

In summary, these numbers are PyPI download totals for Numpy wheels from 30 days prior to 7 August 2019.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('numpy_by_distro_version_id.csv')
df.head(20)

downloads       distro_name distro_version      distro_id
0     4142728  Debian GNU/Linux              9        stretch
1     3485270            Ubuntu          16.04   Xenial Xerus
2     3146568               NaN            NaN            NaN
3     3006325            Ubuntu          16.04         xenial
4     2796493  Amazon Linux AMI        2018.03            NaN
5      928401            Ubuntu          18.04  Bionic Beaver
6      718677            Ubuntu          14.04    Trusty Tahr
7      631047  Debian GNU/Linux             10         buster
8      583995            Ubuntu          18.04         bionic
9      512926  Amazon Linux AMI        2017.09            NaN
10     500226      Amazon Linux              2            NaN
11     426101      CentOS Linux              7           Core
12     352574  Debian GNU/Linux              8         jessie
13     309623  Amazon Linux AMI        2016.03            NaN
14     243489  Amazon Linux AMI        2018.03            NaN
15     214027             macOS        10.14.5            NaN
16      85573             macOS        10.13.6            NaN
17      57191  Amazon Linux AMI        2017.09            NaN
18      53978             macOS        10.14.6            NaN
19      52019  Amazon Linux AMI        2016.09            NaN

In [3]:
by_version = (df.groupby(['distro_name', 'distro_version'])
.sum()
.sort_values('downloads', ascending=False)
             )
by_version.head(10)

downloads
distro_name      distro_version           
Ubuntu           16.04             6491598
Debian GNU/Linux 9                 4142729
Amazon Linux AMI 2018.03           3039995
Ubuntu           18.04             1512398
                 14.04              725595
Debian GNU/Linux 10                 631047
Amazon Linux AMI 2017.09            570117
Amazon Linux     2                  503036
CentOS Linux     7                  426264
Debian GNU/Linux 8                  352574

In [4]:
total_downloads = by_version['downloads'].sum()
total_downloads

19816603

In [5]:
by_version['download_pct'] = by_version['downloads'] / total_downloads * 100
by_version.head(20)

downloads  download_pct
distro_name                     distro_version                         
Ubuntu                          16.04             6491598     32.758379
Debian GNU/Linux                9                 4142729     20.905344
Amazon Linux AMI                2018.03           3039995     15.340646
Ubuntu                          18.04             1512398      7.631974
                                14.04              725595      3.661551
Debian GNU/Linux                10                 631047      3.184436
Amazon Linux AMI                2017.09            570117      2.876966
Amazon Linux                    2                  503036      2.538457
CentOS Linux                    7                  426264      2.151045
Debian GNU/Linux                8                  352574      1.779185
Amazon Linux AMI                2016.03            320243      1.616034
macOS                           10.14.5            214027      1.080039
                                10.13.6             85573      0.431825
Amazon Linux AMI                2017.03             57355      0.289429
macOS                           10.14.6             53978      0.272388
Amazon Linux AMI                2016.09             53823      0.271606
Red Hat Enterprise Linux Server 7.6                 51090      0.257814
Ubuntu                          19.04               42957      0.216773
macOS                           10.13.3             36602      0.184704
                                10.12.6             27750      0.140034

Ubuntu 16.04 and 18.04 may well be polluted with CI runs.

I'm assuming the Debian 9 and Ubuntu 18.04 stats are persisting installs.

* Ubuntu 16.04: Python 3.5, release 4/2016, end of standard support 4/21
* Ubuntu 18.04: Python 3.6, release 4/2018, end of standard support 4/23
* Debian 9: Python 3.5, release 6/2017, EOL ~2020
* Debian 10: Python 3.7, release 7/2019, EOL ~2022

See: [Ubuntu dates](https://wiki.ubuntu.com/Releases), [Debian dates](https://wiki.debian.org/DebianReleases)

Some example criteria:

* Last LTS release as of 2 years ago = Python 3.6 (Ubuntu) 3.7 (Debian)
* Last LTS release as of 3 years ago = Python 3.6 (Ubuntu) 3.5 (Debian), updating to 3.7 (Debian) in 6/2020.

* Debian EOL - Python 3.5 until 2020.